gulpfile.js

```diff
@@ -58,27 +196,62 @@ gulp.task('css:lint', () => {
         }));
 });
 
-gulp.task('static:clean', () => {
-    return del(['static/**', '!static', '!static/.gitignore']);
+/**
+ * Start the browser-sync daemon for local development.
+ */
+gulp.task('browser-sync', ['js:compile', 'css:compile'], () => {
+    const proxyURL = process.env.BS_PROXY_URL || 'localhost:8000';
+    const openBrowser = !(process.env.BS_OPEN_BROWSER === 'false');
+    browserSync({
+        proxy: proxyURL,
+        open: openBrowser,
+        serveStatic: [{
+            route: '/media',
+            dir: 'static_build'
+        }]
+    });
 });
 
-gulp.task('default', () => {
-    gulp.start('media:watch');
+/**
+ * Reload tasks used by `gulp watch`.
+ */
+gulp.task('reload-other', ['assets'], browserSync.reload);
+gulp.task('reload-css', ['css:compile'], browserSync.reload);
+gulp.task('reload-js', ['js:compile'], browserSync.reload);
+gulp.task('reload', browserSync.reload);
 
-    gulp.watch(lintPathsJS).on('change', file => {
-        return gulp.src(file.path)
-            .pipe(eslint())
-            .pipe(eslint.format());
-    });
+// --------------------------
+// DEV/WATCH TASK
+// --------------------------
+gulp.task('watch', ['browser-sync'], () => {
+    gulp.watch([
+        'media/**/*',
+        '!media/css/**/*',
+        '!media/js/**/*'
+    ], ['reload-other']);
 
-    gulp.watch(lintPathsCSS).on('change', file => {
-        return gulp.src(file.path)
-            .pipe(gulpStylelint({
-                failAfterError: false,
-                reporters: [{
-                    formatter: 'string',
-                    console: true
-                }]
-            }));
-    });
+    // --------------------------
+    // watch:css, less, and sass
+    // --------------------------
+    gulp.watch('media/css/**/*', ['reload-css']);
```

Comment:
One other thing I notice is when changing a scss/less file, or a JS file - the watch command seems to be recompiling everything. This means it takes around 6 seconds to see the changes (on docker at least). It might be worth only compiling the files that have changed here?

Reply:
FWIW, running locally/outside of Docker, this takes 10 seconds to re-compile the CSS and reload the browser on my machine. 😞

Reply:
Also FWIW, back when I was on MDN, we were able to avoid re-compiling everything because some smart person (probs @stephaniehobson) organized the CSS so that all includes/non page CSS files were kept in a separate folder. This meant the re-compile only had to target the "higher level" `.css/.scss/.less` files (as recompiling them would infer a recompile of any `.css/.scss/.less` files they `@import`'ed).

Even if we did feel like re-architecting our CSS folder structure, the sheer size of moz.org may prevent it from being a huge boost.

I'm not saying we should stick with Django Pipeline, but one nice thing it granted was only recompiling the files of the one page being loaded.

Reply:
I think it should be possible to only compile the changed file here. Even if we can’t figure out auto compilation of imported files right away, it still might be preferable to endure rather than a really long compile time.

Reply:
IIRC, the issue with only compiling the changed file is if that changed file is a library/include. The files including that library/include wouldn't get recompiled, so you wouldn't see the change reflected on any pages until the including files were also modified/recompiled. (I think this is basically what pmac was getting at above.)

Reply:
Yup, I get the problem. I seem to remember we had this same issue a while back, before we switched to Django Pipeline.

Reply:
There seem to be some solutions out there possibly. This is the first one I found but there seems to be a few others as well: https://github.com/berstend/gulp-sass-inheritance

Reply:
@jpetto I'm actually already only compiling the files used in pages, since I'm only compiling those less and sass files referenced in bundles.

Reply:
It's just that we have kindof a ton of those :)

Reply:
I also had it working a bit better before adding bundling since I could just recompile less or sass individually, but now I have to do the bundling step as well before the page will work again. I could potentially figure out to which bundle(s) a file belonged and just rebuild it. We could also just document that changing a file not directly included in a bundle will require the user to call `gulp css:compile` for the whole lot. Better?

Reply:
If working on an include/library file, the need to run any gulp task with *every file change* would be, IMO, a deal-breaking inconvenience. CSS dev is often many rapid changes and reloads. An extra command in that chain would be really expensive.

If we want to stick to compiling pre-page request, I think we need to find some lib like Alex linked above, or do something like Paul mentioned above, where all and only the set of files impacted by the actual file changed are recompiled.

Reply:
I'm going to explore these things alex brought up, mostly the `gulp-cached` and then the `gulp-sass-inheritance`. I think `gulp-cached` will get us a long way actually.